Adding a schema file when creating a snapshot

amnonh opened the pull request:
To use a snapshot we need a schema file that is similar to the result of running cql DESCRIBE command.

The DESCRIBE is implemented in the cql driver so the functionality needs to be re-implemented inside scylla.

This series adds a describe method to the schema file and use it when doing a snapshot.

There are different approach of how to handle materialize views and secondary indexes. 

This implementation creates each schema.cql file in its own relevant directory, so the schema for materializing view, for example, will be placed in the snapshot directory of the table of that view.

Fixes #4192

denesb commented on view_info.hh:
```diff
@@ -58,6 +58,10 @@ public:
+    /*!
+     * \brief returns true if this view is backing a secondary index
+     */
+    bool is_secondary_index() const;
```
This is a layering violation, a materialized view shouldn't know about different things it is used for as an infrastructure. In fact we used to have this method and @psarna removed it in an effort to fix this layering violation.

amnonh replied:
so you remove the method, but still do that in the code,
see the code in: cql3::statements::select_statement& view_info::select_statement()

I'll place the logic somewhere else

psarna replied:
The code in `view_info::select_statement` is a necessary evil due to backward compatibility, it's marked with this comment:
```cpp
// FIXME(sarna): legacy code, should be removed after "computed_columns" feature is guaranteed
// to be available on every node. Then, we won't need to check if this view is backing a secondary index.
```
It would be best to avoid checking if if the view backs an index in the view code itself, but instead on an upper layer (e.g. a code that iterates over all views of a table could exclude ones backed by an index, because a table knows its indexes and their accompanying views)

amnonh replied:
@psarna you're right but that is not the case I'm facing, I'm writing a describe method in a schema (similar to a cql DESCRIBE command) and if the schema in hand is a view I need to know how to treat it.

Anyway, as I replied to @denesb I'll move the same hack that is used here to somewhere else for my need.

denesb commented on schema.cc:
```diff
@@ -656,6 +656,135 @@ std::ostream& operator<<(std::ostream& os, const schema& s) {
+
+    if (is_view()) {
+        if (view_info()->is_secondary_index()) {
+            os << "INDEX " << cf_name().substr(0, cf_name().size() - 6) << " ON " << ks_name() << "." << view_info()->base_name() << "(";
```
What is this `- 6`?

amnonh replied:
remove the _index prefix I'll make it clearer

denesb commented:
Please write unit tests that verify that parsing back the output of `schema::describe()` results in a schema equivalent to the original one. You can use `tests/random_schema.hh` to generate random schemas but you should also generate random (but valid) schema properties as well (random_schema doesn't do that). You can then generate a materialized view and secondary index for the random schema and check that they describe correctly as well.

amnonh commented:
ok, naturally I checked it with the different options, but I'll add tests for it as well.

avikivity commented:
What does "A schema file when doing a snapshot" mean? I can't parse it.

amnonh commented:
I'm adding some tests and I'll update the description when I'm pushing the next version

amnonh commented:
@denesb I've forced push V2.

Main changes, I remove the is_index from the view layer and mimic that functionality where it is needed.

I've added some basic testing for a table, an index and materialized view.

psarna commented on schema.cc:
```diff
@@ -656,6 +658,145 @@ std::ostream& operator<<(std::ostream& os, const schema& s) {
+    return name.substr(0, name.size() - 6); // remove the _index suffix from an index name;
+}
+
+static bool is_secondery_index(const utils::UUID& id, const schema& s) {
```
there's a typo here :) is_second**a**ry_index

Also, the commit message  has `schmea` instead of `schema`

amnonh replied:
@psarna thanks, I've forced push V3 with a fix

slivne commented:
@psarna please comment / approve the last version

psarna commented on tests/cql_query_test.cc:
```diff
@@ -4261,3 +4261,135 @@ SEASTAR_TEST_CASE(test_rf_expand) {
+                "    AND memtable_flush_period_in_ms = 0\n"
+                "    AND min_index_interval = 128\n"
+                "    AND read_repair_chance = 0\n"
+                "    AND speculative_retry = '99.0PERCENTILE';\n");
```
What about comparing the schemas with `cmp_schema` for this case?

Also, I think this big description strings should be moved to a string variable and used both during table creation and here. Reasons for that:
1. You start by creating the table without specifying all these configuration variables like `bloom_filter_fp_chance`, and then compare the results *with* these values - which means that every single change in the default value will also unnecessarily invalidate this test case.
2. It will simply result in less code duplication.

amnonh replied:
@psarna I've forced push a new version, switching to a single string to create a table and to compare the generated schema to remove some of the tests.

Because  it's the same string, we know it works (or there wouldn't be a schema)

psarna commented on tests/cql_query_test.cc:
```diff
@@ -4261,3 +4261,135 @@ SEASTAR_TEST_CASE(test_rf_expand) {
+
+        std::ostringstream ss1;
+        schema->describe(ss1);
+        BOOST_CHECK_EQUAL(ss1.str(), "CREATE INDEX cf_index ON ks.cf(col2);\n");
```
again, shouldn't schemas be compared again with `cmp_schema` after the second change?

psarna commented on tests/cql_query_test.cc:
```diff
@@ -4261,3 +4261,135 @@ SEASTAR_TEST_CASE(test_rf_expand) {
+
+static void cmp_schema(schema_ptr s1, schema_ptr s2) {
+    BOOST_CHECK_EQUAL(s1->ks_name(), s2->ks_name());
+    BOOST_CHECK_EQUAL(s1->cf_name(), s2->cf_name());
```
One more really easy check to add would be to check if the schemas have the same column names (and perhaps types). There are multiple ways of listing column names from the schema, one of them is `columns_by_name()`, which gives you a map where keys are ordered column names - easy to compare.

I'm also fine with leaving it as a todo for the future.

slivne commented:
@amnonh ping

slivne commented:
@amnonh ping

psarna commented on table.cc:
```diff
@@ -1747,7 +1764,7 @@ future<> table::snapshot(sstring name) {
                     table_names.insert(std::move(rf));
                 }
-                return smp::submit_to(shard, [requester = engine().cpu_id(), jsondir = std::move(jsondir),
+                return smp::submit_to(shard, [requester = engine().cpu_id(), jsondir = std::move(jsondir),this,
```
I only found one super-tiny style mistake - please put a space between the comma and 'this' - I can push the series after this.

amnonh replied:
I have forced push an updated version with the missing space

psarna commented:
queued to /next

psarna commented:
Dequeued: a unit test failed (database_test).

Please fix and include an information that unit tests were run, e.g.:
```
Tests: unit(dev)
```
Also, note that I'm starting vacation in an hour, so this pull request likely needs attention of another maintainer :)

amnonh commented:
@psarna Enjoy your vacation.

Found the issue, the problem is with a different test that needs to be changed

amnonh commented:
I have forced push a new version, get_snapshot_details in table.cc should ignore the schema.cql file or the relevant test will fail.

slivne commented:
@nyh can you please help ... sarna is on vacation

@amnonh please rebase

nyh commented:
On Sun, Dec 22, 2019 at 12:36 PM Shlomi Livne <notifications@github.com>
wrote:

> @nyh <https://github.com/nyh> can you please help ... sarna is on vacation
>
> @amnonh <https://github.com/amnonh> please rebase
>

I'll review it now, but to your long list of problems in our long review
process you can add another one:
When you switch reviewer in the middle of a long review, the new reviewer
may have different opinions and care about new things that the original
reviewer never cared about, resulting in one of two options, I don't know
which of them is worse:
1. The new reviewer may demand new things and slow the patch's commit even
more.
2. The new reviewer may commit the patch contrary to his personal opinions,
because the previous reviewer as almost fine with this patch.

nyh commented on table.cc:
```diff
@@ -1713,6 +1713,23 @@ seal_snapshot(sstring jsondir) {
+    auto json = ss.str();
+    auto jsonfile = dir + "/schema.cql";
+    return open_checked_file_dma(general_disk_error_handler, jsonfile, open_flags::wo | open_flags::create | open_flags::truncate).then([json](file f) {
+        return do_with(make_file_output_stream(std::move(f)), [json] (output_stream<char>& out) {
```
I think you can move json (which isn't json...) instead of copying it?

nyh commented on tests/cql_query_test.cc:
```diff
@@ -4261,3 +4261,106 @@ SEASTAR_TEST_CASE(test_rf_expand) {
+        e.execute_cql("DROP TABLE ks.cf").get();
+        e.execute_cql(ss.str()).get();
+
+        std::string create_table_with_compund_key = "CREATE  TABLE ks.cf1 (\n"
```
There are other variants you should try, e.g., multiple partition key columns (which is a special case you have special code with parentheses for), and importantly - uppercase table names, quoted column names, etc.
To reduce the amount of duplicate code, you can have an array of strings with just the different part (appending at the end the common part with all those silly parameters), and do the test with a loop instead of duplicate code.

amnonh replied:
sounds good, I'll squeeze it into a single test with a loop

nyh commented on tests/cql_query_test.cc:
```diff
@@ -4261,3 +4261,106 @@ SEASTAR_TEST_CASE(test_rf_expand) {
+        std::ostringstream ss1;
+        schema->describe(ss1);
+        BOOST_CHECK_EQUAL(ss1.str(), create_index);
+    });
```
Please test at the end of this test that describe() of "cf", (not "cf_view" or "cf_index_index"). I don't think it's obvious whether describe of a table *with* materialized views should list them or not - but since in the code we decided NOT to list them, the test should verify this: describe() of cf, despite it having a materialized view and an index, should contain CQL to just create cf, not its view.

amnonh replied:
good point, I'll add a check for that as well

amnonh commented on schema.cc:
```diff
@@ -656,6 +658,145 @@ std::ostream& operator<<(std::ostream& os, const schema& s) {
+static bool is_secondary_index(const utils::UUID& id, const schema& s) {
+    return  service::get_local_storage_service().db().local().find_column_family(id).get_index_manager().is_global_index(s);
+}
+std::ostream& schema::describe(std::ostream& os) const {
```
Sure, I'll add some explenation

amnonh commented on schema.cc:
```diff
@@ -656,6 +658,145 @@ std::ostream& operator<<(std::ostream& os, const schema& s) {
+
+    if (is_view()) {
+        if (is_secondary_index(view_info()->base_id(), *this)) {
+            os << "INDEX " << remove_index_suffix(cf_name()) << " ON " << ks_name() << "." << view_info()->base_name() << "(";
```
it's based on a comment from the previous committer it removes the _index suffix from the index name, if you didn't find it clear, I'll think of a different name for it.

nyh replied:
The problem is that by default, there are two "index suffixes" see issue #4146 - and by this name it's impossible to guess which of these two "suffixes" the function removes.
Maybe the name is ok if it includes a clear comment on what it does, but I don't see the definition of this function - again - where is it? a separate patch? This function should be in `index/secondary_index_manager.cc`, right next to `index_table_name()`. You could call your new function `from_index_table_name()` to make it clearer what it does, and that it is the reverse of `index_table_name()`.

amnonh commented on schema.cc:
```diff
@@ -656,6 +658,145 @@ std::ostream& operator<<(std::ostream& os, const schema& s) {
+    if (is_view()) {
+        if (is_secondary_index(view_info()->base_id(), *this)) {
+            os << "INDEX " << remove_index_suffix(cf_name()) << " ON " << ks_name() << "." << view_info()->base_name() << "(";
+            for (auto& pk : partition_key_columns()) {
```
I'll add one

nyh replied:
Did you? You have to. Otherwise, your code will seem to work until someone happens to use it with local secondary indexes, and things just silently break :-(
We *can* commit code without the local-secondary-index part and fix this bug later, hoping nobody uses local secondary indexes (?) but that's not a technical decision but rather a decision on how quickly we need a almost-working fix in. I have no input on that. Maybe @slivne does?

amnonh commented on schema.cc:
```diff
@@ -656,6 +658,145 @@ std::ostream& operator<<(std::ostream& os, const schema& s) {
+            os <<");\n";
+            return os;
+        } else {
+            os << "MATERIALIZED VIEW " << ks_name() << "." << cf_name() << " AS\n";
```
it should work, but I'll verify.

nyh replied:
I don't see how this can work without being fixed. If cf_name is FooBar (capitalized like this), the only correct way to create is with the CQL command:
```
CREATE TABLE "FooBar"
```
Like this - with double quotes. Without the double quotes, it will just create a table called foobar - and this is not the table you wanted to create. Fixing this is trivial - you can use the existing function, `cql3::util::maybe_quote(sstring(cf_name))` to do this CQL quoting if necessary.

amnonh replied:
I already tested it again and validate that there is an issue with the table name. The next version will take care of that and I have tests that validate that it's correct.

While the parameters are quoted correctly, that table/ks wheren't (but it will be ok)

nyh replied:
Right. The column names you did quote correctly - just not the table names (the index name, view name, and table name, can all have this issue).

amnonh commented on schema.cc:
```diff
@@ -656,6 +658,145 @@ std::ostream& operator<<(std::ostream& os, const schema& s) {
+            os << "\n    WHERE " << view_info()->where_clause();
+        }
+    } else {
+        os << " TABLE " << ks_name() << "." << cf_name() << " (";
```
I'll add test for it

amnonh commented on schema.cc:
```diff
@@ -656,6 +658,145 @@ std::ostream& operator<<(std::ostream& os, const schema& s) {
+        os << pk.name_as_cql_string();
+    }
+    os << ")";
+    if (is_view()) {
```
I'll check

amnonh commented on table.cc:
```diff
@@ -1713,6 +1713,23 @@ seal_snapshot(sstring jsondir) {
+future<> table::write_schema(sstring dir) const {
+    std::ostringstream ss;
+    this->schema()->describe(ss);
+    auto json = ss.str();
```
yes, you're right, I'll change it

amnonh commented:
@nyh Thanks for the review. I'm just forced pushed a new version. Capitalization of table and views are now handled correctly and is tested. 
Additional changes, I moved the function that returns the index name to secondary_index_manager and rename it.

I also cleaned the variables that used json in their names.

slivne commented:
@nyh ping

nyh commented on index/secondary_index_manager.cc:
```diff
@@ -100,6 +100,10 @@ sstring index_table_name(const sstring& index_name) {
 }
 
+sstring index_name_from_table_name(const sstring& table_name) {
+    return table_name.substr(0, table_name.size() - 6); // remove the _index suffix from an index name;
```
It would have been better to check and throw an exception if the table_name doesn't end with "_index". It's a very cheap test (this is a very rarely called function) and can help serious bugs in using this function.

nyh commented on schema.hh:
```diff
@@ -909,6 +909,15 @@ public:
+     * CQL DESCRIBE is implemented at the driver level. This method mimic that functionality
+     * inside Scylla.
+     *
+     * It should be possible to use the method result to re-create the schema.
```
Please be more explicit about what this function does schemas which **have** a view or an index, and for schemas which **are** a view or index itself. For example something like:
This function returns a "CREATE TABLE", "CREATE MATERIALIZED VIEW" or "CREATE INDEX" command depending on whether the given schema is a base table, a materialized view or a materialized view created for a secondary index. If the given schema is a base table, only the command to create the table itself is returned - its materialized views or indexes, if any, need to be created separately.

amnonh replied:
ok

nyh commented on schema.cc:
```diff
@@ -638,6 +640,141 @@ std::ostream& operator<<(std::ostream& os, const schema& s) {
+}
+
+static bool is_secondary_index(const utils::UUID& id, const schema& s) {
+    return  service::get_local_storage_service().db().local().find_column_family(id).get_index_manager().is_global_index(s);
```
You left out local indexes. Your patch is incomplete without recognizing those, and also recreating them with the proper CQL command. And of course, you need to add a test case for them, which will show you you missed them. I already pointed out this in my previous round of review - see my other comment below which remains.

slivne commented:
we need secondary indexes to work if anyone is in doubt

nyh commented on database.hh:
```diff
@@ -827,6 +827,10 @@ public:
     future<std::unordered_map<sstring, snapshot_details>> get_snapshot_details();
 
+    /*!
+     * \brief write the table schema to a schema file at the given directory
```
Would have been nice to give more context in this comment... That this function has to do with snapshots (this "schema file" isn't something we use normally in normal database operation), and the "schema file" is really a CQL command to recreate the schema.
I think it would also be better to rename this function **write_snapshot_schema** or similar, to make clearer this function is just used by snapshots (if "snapshot" is the right word... Or is it "backup"?).

amnonh replied:
ok, I'll take your view, though it's a valid schema, if anyone would need the schema description (i.e pretty print instead of the redirection operator) it would be useful.

nyh replied:
So maybe write_schema_as_cql() or something like that?
To emphasize that it writes CQL, it's not the normal function we use to persist a schema to disk or something.

amnonh replied:
I like that better

nyh commented on schema.cc:
```diff
@@ -638,6 +640,141 @@ std::ostream& operator<<(std::ostream& os, const schema& s) {
+
+std::ostream& column_definition_as_cql_key(std::ostream& os, const column_definition & cd) {
+    os << cd.name_as_cql_string();
+    os << " " << cd.type->cql3_type_name();
```
Please consider: What happens when UDT are used? Do we need to save the UDT definition in the CQL file as well? Does this cql3_type_name() method (I'm not familiar with it) work correctly also with UDTs and also with UDTs with quoted names (I believe this is possible, but didn't check...)
If UDTs don't work, we can leave it as a TODO, but we should throw an exception from describe() rather than silently doing the wrong thing.

amnonh replied:
UDT and UDF are defined at the keyspace level. We assume that the keyspace is defined correctly before creating the table.

you are right that I should test it as well

nyh replied:
I keep asking and don't understand the response - is it fine that there is no maybe_quote on the cql3_type_name?

nyh commented on test/boost/cql_query_test.cc:
```diff
@@ -4313,3 +4315,198 @@ SEASTAR_TEST_CASE(test_rf_expand) {
+    return do_with_cql_env_thread([] (cql_test_env& e) {
+        std::unordered_map<std::string, std::string> cql_create_tables {
+            {"cf", "CREATE TABLE ks.cf (\n"
+            "    pk blob,\n"
```
nitpick: the indentation of this "cf" is different from the following ones.

amnonh replied:
ok

nyh commented on test/boost/cql_query_test.cc:
```diff
@@ -4313,3 +4315,198 @@ SEASTAR_TEST_CASE(test_rf_expand) {
+                    "    AND speculative_retry = '99.0PERCENTILE';\n"
+            },
+            {"CF2", "CREATE TABLE ks.\"CF2\" (\n"
+                    "    pk blob,\n"
```
It seems in none of these tests you tried a quoted partition-key column (only clustering key), and neither did you try quoted regular column names.

nyh commented on test/boost/cql_query_test.cc:
```diff
@@ -4313,3 +4315,198 @@ SEASTAR_TEST_CASE(test_rf_expand) {
+                    "    AND read_repair_chance = 0\n"
+                    "    AND speculative_retry = '99.0PERCENTILE';\n"
+            },
+            {"cf4", "CREATE TABLE ks.cf4 (\n"
```
what does cf4 test which cf2 didn't?

amnonh replied:
no clustering key

nyh replied:
Oh, I see you have a column called "CK" but it's no longer a clustering key :-) Confusing, but right. Ok.

nyh commented:
> we need secondary indexes to work if anyone is in doubt

I don't think that regular secondary indexes work with this patch, I was worried about the newer "local secondary index" feature and whether this code knows how to describe its schema. Maybe it does and I just don't understand how? But we should at least have a test case for this case.

amnonh commented:
@nyh I have forced push a new version, I hope everything is covered.
Besides the changes, there is also a test for a local index, a combination of compound key and a test with User define type.

nyh commented:
Thanks. I'll review - and hopefully push - now.

nyh commented on test/boost/cql_query_test.cc:
```diff
@@ -4313,3 +4315,207 @@ SEASTAR_TEST_CASE(test_rf_expand) {
+              "    AND min_index_interval = 128\n"
+              "    AND read_repair_chance = 0\n"
+              "    AND speculative_retry = '99.0PERCENTILE';\n"},
+          {"cf_index_index", "CREATE INDEX cf_index ON ks.cf(col2);"},
```
Would also be good to check three more indexes: ks.cf(pk), ks.cf(pk1) and ks.cf(ck) - because it's possible to index also a key column (I think your code is correct in this case - I don't suspect a bug - but it wouldn't hurt to check, and it should be very easy with your code that can easily add more tests in this looo).

amnonh commented on test/boost/cql_query_test.cc:
```diff
@@ -4313,3 +4315,207 @@ SEASTAR_TEST_CASE(test_rf_expand) {
+            }
+
+        };
+        e.execute_cql("CREATE TYPE phone ("
```
From what I've tested, if the type is quoted-capitalized the type will be created, but the create table will fail because it will still look for the lower case name.

nyh replied:
So, the code is wrong, right (I left you a comment exactly where I think the code is wrong)? So why not fix it? And why not add the test here to confirm the code has been fixed?

amnonh commented:
Thanks @nyh that was my last forced-push for 2019, let 2020 brings joy to all!

amnonh commented:
On Tuesday, December 31, 2019, nyh <notifications@github.com> wrote:

> *@nyh* requested changes on this pull request.
> ------------------------------
>
> In schema.cc
> <https://github.com/scylladb/scylla/pull/5294#discussion_r362255690>:
>
> > +std::ostream& map_as_cql_param(std::ostream& os, const std::map<sstring, sstring>& map, bool first = true) {
> +    for (auto i: map) {
> +        if (first) {
> +            first = false;
> +        } else {
> +            os << ",";
> +        }
> +        os << "'" << i.first << "': '" << i.second << "'";
> +    }
> +
> +    return os;
> +}
> +
> +std::ostream& column_definition_as_cql_key(std::ostream& os, const column_definition & cd) {
> +    os << cd.name_as_cql_string();
> +    os << " " << cd.type->cql3_type_name();
>
> I keep asking and don't understand the response - is it fine that there is
> no maybe_quote on the cql3_type_name?
>
I missed that you were asking about it, it's a wrapper that call maybe
quote internally.

------------------------------
>
> In test/boost/cql_query_test.cc
> <https://github.com/scylladb/scylla/pull/5294#discussion_r362255954>:
>
> > +                "    AND comment = ''\n"
> +                "    AND compaction = {'class': 'SizeTieredCompactionStrategy'}\n"
> +                "    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}\n"
> +                "    AND crc_check_chance = 1\n"
> +                "    AND dclocal_read_repair_chance = 0.2\n"
> +                "    AND default_time_to_live = 0\n"
> +                "    AND gc_grace_seconds = 954000\n"
> +                "    AND max_index_interval = 3048\n"
> +                "    AND memtable_flush_period_in_ms = 10\n"
> +                "    AND min_index_interval = 128\n"
> +                "    AND read_repair_chance = 0\n"
> +                "    AND speculative_retry = '99.0PERCENTILE';\n"
> +            }
> +
> +        };
> +        e.execute_cql("CREATE TYPE phone ("
>
> So, the code is wrong, right (I left you a comment exactly where I think
> the code is wrong)? So why not fix it? And why not add the test here to
> confirm the code has been fixed?
>
Are you talking about the user define type?
If so, I responded on the previous patch, I've tested with quoted name and
it does not supported in a kind of weird way.

If it's something else, then Im not sure what you are referring to.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/5294?email_source=notifications&email_token=AAQFDP73Q63SUGDIXORN2QLQ3OBW3A5CNFSM4JOPBTG2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQOKU2Y#pullrequestreview-337422955>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQFDP47WYGMA6HFZDDI3MDQ3OBW3ANCNFSM4JOPBTGQ>
> .
>

nyh commented:
On Wed, Jan 1, 2020 at 12:47 AM Amnon Heiman <notifications@github.com>
wrote:

> > + os << " " << cd.type->cql3_type_name();
> >
> > I keep asking and don't understand the response - is it fine that there
> is
> > no maybe_quote on the cql3_type_name?
> >
> I missed that you were asking about it, it's a wrapper that call maybe
> quote internally.
>

Excellent, I didn't find where this is implemented so I was worried it is
missing a maybe_quote.
Especially since you said that types with quoted names didn't work
correctly in your test.


> > + e.execute_cql("CREATE TYPE phone ("
> >
> > So, the code is wrong, right (I left you a comment exactly where I think
> > the code is wrong)? So why not fix it? And why not add the test here to
> > confirm the code has been fixed?
> >
> Are you talking about the user define type?
>

Yes, exactly. You can have exactly the same test with the type name "pHoNe"
and
it should work too (unless I'm misunderstanding something in how type names
work
in Cassandra or Scylla).

If so, I responded on the previous patch, I've tested with quoted name and
> it does not supported in a kind of weird way.
>

"not supported in a weird way", means it doesn't work? Is it *supposed* to
work (works in Cassandra)? If so, why not fix it?

If it's a completely unrelated bug in Scylla with support for quoted type
names, please open a separate issue about this.
But is it completely unrelated and just a bug in describe()? I don't
understand how. In any case, if it's easy to fix, better just fix it - and
add a test for this case.

amnonh commented:
@nyh My first forced-push to 2020 added the additional index tests

nyh commented:
Committed to next. Mazal tov ;-)

avikivity commented:
Dequeueing, fails cql_query_test:

```
ecbabe2f42ae269ed8b8f5dc860cf01081f8b7de is the first bad commit
commit ecbabe2f42ae269ed8b8f5dc860cf01081f8b7de
Author: Amnon Heiman <amnon@scylladb.com>
Date:   Mon Nov 18 09:03:22 2019 +0200

    database: add schema.cql file when creating a snapshot
    
    When creating a snapshot we need to add a schema.cql file in the
    snapshot directory that describes the table in that snapshot.
    
    This patch adds the file using the schema describe method.
    
    get_snapshot_details was modified to ignore the schema.cql file.
    
    Fixes #4192
    
    Signed-off-by: Amnon Heiman <amnon@scylladb.com>

 database.hh | 18 ++++++++++++++++++
 table.cc    | 30 +++++++++++++++++++++++++-----
 2 files changed, 43 insertions(+), 5 deletions(-)
```

amnonh commented:
@nyh 
I have forced push a version that prints an error when writing a schema but does not stop the snapshot.
cql_query_test now passes

nyh commented:
@amnonh it would have been nice to put here (for lack of a better place... I'm not a fan of these pull requests...) a description of what you fixed in this version (although I know :-)). I'll review and commit.

nyh commented:
Hmm, I now realized that "prints an error when writing a schema but does not stop the snapshot" was supposed to be a description of what you think... But I'm not sure how anyone except the two of us will know what this means and which problem in the test it fixed.

nyh commented:
@amnonh unfortunately your branch does not apply cleanly to "next". Please rebase.

slivne commented:
@amnonh ping

amnonh commented:
I have forced push a newer version this morning, but I see that there is already another conflict

nyh commented:
It seems that today, everybody is sending patches to cql_query_test.cc, and all of these patches want to add a new test at the end, so all of them conflict :-( Whoever rebases their patches more quickly, has a better chance to be committed without yet another rebase - so - the race is on! :-)

amnonh commented:
2AM is as good a time to force push pull-rebase

nyh commented:
Committed to next. Hopefully 10th time is the charm :-)

amnonh commented:
Praying for the old gods and the new

avikivity commented:
Dequeued due to dtest failures.

amnonh commented:
I've forced push a new version, but there is also a change in dtest that need to be merged first:
https://github.com/scylladb/scylla-dtest/pull/1247

avikivity commented:
What are the changes in the new version?

nyh commented:
@amnonh I see your dtest patch was already merged into scylla-dtest next. @fruch what is the process of how it reaches scylla-dtest master?

But you said you forced-pushed a new version of the Scylla patch as well. Why? What changed? (@avikivity asked this as well).

fruch commented:
The dtest would be in master once next is passing

amnonh commented:
I've updated the message following @avikivity comment to state that table::manifest_json_filter now filter the schema.cql file.

I see that there is now a conflict in the cql_query_test as well, I'll rebase

nyh commented:
Yes, I saw your changes to dtest, but you also said that you forced-push a version of Scylla, here, so I was wondering why. Was it just a rebase? Any other change?

amnonh commented:
@nyh there is an additional change:
table::manifest_json_filter now filter the schema.cql file.

nyh commented:
Ok, I don't know what to do now but commit the new version and pray you fixed the tests.
I'll do that - minus the praying ;-)

nyh commented:
Well, I can't commit it yet, actually, because your dtest patch **still** hasn't reached master (https://github.com/scylladb/scylla-dtest/commits/master), so if I commit this patch, it will fail the dtest again. This is so frustrating :-(  If I forget to commit your series tomorrow, please remind me again.

amnonh commented:
@nyh I forced-push a rebase of the series

nyh commented:
Pushed to next. Please consider praying to a more powerful god this time :-)
